src/core/Class.js

```diff
@@ -4,114 +4,112 @@ import * as Util from './Util';
 // @aka L.Class
 
 // @section
-// @uninheritable
 
 // Thanks to John Resig and Dean Edwards for inspiration!
 
-export function Class() {}
-
-Class.extend = function (props) {
-
-	// @function extend(props: Object): Function
-	// [Extends the current class](#class-inheritance) given the properties to be included.
-	// Returns a Javascript function that is a class constructor (to be called with `new`).
-	const NewClass = function (...args) {
+export class Class {
+	_initHooksCalled = false;
 
+	constructor(...args) {
 		Util.setOptions(this);
 
 		// call the constructor
 		if (this.initialize) {
-			this.initialize.apply(this, args);
+			this.initialize(...args);
 		}
 
 		// call all constructor hooks
 		this.callInitHooks();
-	};
+	}
 
-	const parentProto = this.prototype;
+	callInitHooks() {
+		if (this._initHooksCalled) { return; }
 
-	const proto = Object.create(parentProto);
-	proto.constructor = NewClass;
+		// collect all prototypes in chain
+		const protos = [];
+		let proto = Object.getPrototypeOf(this);
 
-	NewClass.prototype = proto;
+		while (proto !== null) {
+			protos.push(proto);
+			proto = Object.getPrototypeOf(proto);
+		}
+
+		// reverse so the parent prototype is first
+		protos.reverse();
 
-	// inherit parent's statics
-	for (const i in this) {
-		if (Object.hasOwn(this, i) && i !== 'prototype') {
-			NewClass[i] = this[i];
+		// call init hooks on each prototype
+		for (const proto of protos) {
+			const initHooks = proto._initHooks ?? [];
+			for (const hook of initHooks) { hook.call(this); }
 		}
-	}
 
-	// mix static properties into the class
-	if (props.statics) {
-		Util.extend(NewClass, props.statics);
+		this._initHooksCalled = true;
 	}
 
-	// mix includes into the prototype
-	if (props.includes) {
-		Util.extend.apply(null, [proto].concat(props.includes));
-	}
+	// @function extend(props: Object): Function
+	// [Extends the current class](#class-inheritance) given the properties to be included.
+	// Returns a Javascript function that is a class constructor (to be called with `new`).
+	static extend({statics, includes, ...props}) {

```

Comment:
Should be the new default way `class Parent extends L.Class` or `L.Class.extend()`? 
If they are resulting in the exact same, then I would say we mark this function als deprecated. 
But for me it looks like they are different. With `.extend()` we apply `statics` and `includes` over `extends L.Class` we didn't. Or do I oversee something?

Reply:
I would like the default to be `class Parent extends L.Class`, with `.extend()` left just for compatibility.

> With `.extend()` we apply `statics` and `includes` over `extends L.Class` we didn't. Or do I oversee something?

That's the current behaviour if I understand the code correctly. I think it'd be possible to apply `statics` and `includes` and `options` in the `constructor()` as well.

Reply:
The new default should become the ECMAScript syntax. However I have yet to find a way to apply the `statics`, `includes` and `options` as they all modify the prototype. This kind of modification should happen when the class is defined (e.g. `Class.extends()`), and not when it is instantiated.

Not sure if we can land this in a backwards compatible manner, but perhaps we could provide static methods on `Class` to emulate these APIs. We already have `Class.include()` so we could create a `Class.addOptions()` and `Class.addStatics()` method.

Any thoughts?